### PR TITLE
NIFI-4710 Kerberos support for user auth in Docker instances

### PR DESCRIPTION
**Pull request by SarthakSahu**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by SarthakSahu** [comment]
Is anyone having bandwidth to review this.

**Comment by SarthakSahu** [comment]
@apiri do you  have any free time to review this.

**Comment by apiri** [comment]
hey @SarthakSahu, will put this on my list of things to scope out.  sorry for the delay but really appreciate you taking this task on.  Going to list a couple of other folks that might have knowledge on the topic in case they have any cycles.
@pepov @jtstorck

**Comment by pepov** [comment]
hey, I have some experience with the kerberos setup in kubernetes and I can check this, but I need a day or too

**Comment by SarthakSahu** [comment]
@pepov This Jira is only for user authentication through Kerberos. So kerebros zookeeper support has not provided here.

**Comment by pepov** [comment]
I didn't mean to involve Zookeeper in this.

**Comment by pepov** [comment]
To clarify, I didn't want to ask to extend the existing docker-compose file but to create one that can be used to test the user authentication through Kerberos. However it's not necessarily a problem if you cannot do that, but please provide the steps to be taken to setup the necessary test environment (don't need to get into details about the certificate part).

My concern is that I don't understand how this can work without a keytab.

**Comment by SarthakSahu** [comment]
@pepov ok. 
1. Service Principal  and service keytab required where as nifi needs authenticate as client. So when nifi needs to authenticate as client to kerborized zookeeper service keytab and principal is needed.

2. For user authentication nifi will authenticate users with help of KDC server.
For example: A "pepov" user  wants to authenticate with nifi.
We need to create a principal in KDC server as pepov@EXAMPLE.COM using below command.
Note: we have not used -randkey option. Instead we provide password to the created princpal
kadmin:  addprinc pepov
WARNING: no policy specified for pepov@EXAMPLE.COM; defaulting to no policy
Enter password for principal "pepov@EXAMPLE.COM":
Re-enter password for principal "pepov@EXAMPLE.COM":
Principal "pepov@EXAMPLE.COM" created.
Now we have to provide the username(pepov) and password in the login page to access nifi.

Hope this clarify the query.

**Comment by pepov** [comment]
Yes, so I was confused because I supposed single sign on is the default way we wan't to configure kerberos authentication: https://nifi.apache.org/docs/nifi-docs/html/administration-guide.html#kerberos_service

But I just realized the most basic setup you provided authenticates the user against kerberos on behalf of the user: https://nifi.apache.org/docs/nifi-docs/html/administration-guide.html#kerberos_login_identity_provider

I'm still new to this but I suppose my confusion comes from the fact that Kerberos is basically designed to provide single sign on authentication where users authenticate themselves against the KDC and got a ticket that can be used to authenticate to the service (NiFi in this case) without the need from the service to communicate with the KDC. But this process requires the service to have an SPN created in Kerberos and the keytab has to be available. I beleive this is what NiFi calls the SPNEGO service principal and keytab (`nifi.kerberos.spnego.principal*`, `nifi.kerberos.spnego.keytab.location*`) and this is what I was referring to by service principal and keytab: https://nifi.apache.org/docs/nifi-docs/html/administration-guide.html#kerberos_properties

What you mentioned is the other SPN and keytab, referred to as `nifi.kerberos.service.principal*`, `nifi.kerberos.service.keytab.location*` which is used by NiFi to authenticate to other systems as you just mentioned, but this is not what I was referring to.

**Comment by pepov** [comment]
All in all if I was right in the previous comment I would highlight that this is the convenience method and is not the most secure way to authenticate against Kerberos since it requires the users to give out their passwords (even if it's over a TLS encrypted connection) and NiFi needs direct access to the KDC. Both of these can be avoided using the SPNEGO setup but this changeset does not implement that.

@bbende please confirm or correct me here

**Comment by bbende** [comment]
@pepov your description of the various kerberos options is correct, and the SPNEGO approach is definitely the traditional way kerberos is intended to work. The direct login is definitely more convenient in that you don't have to kinit or configure your browser, however it does somewhat defeat the purpose of kerberos by having the user provide their credentials directly to NiFi.

**Comment by alopresto** [comment]
To echo what @bbende said, SPNEGO is far preferred over the form field for Kerberos username and password, that feature was provided solely as a fallback system for legacy browsers that did not support GSS-API SSO, and there have been requests to remove or disable that functionality completely. This addition should support the SPNEGO feature as a first class feature.

**Comment by pepov** [comment]
@SarthakSahu we didn't mean to discourage you, I beleive it would be great to have the SPNEGO setup working and it's not far from the proposed changeset at all. When you create the user principal you can also create another one (the SPN) for NiFi as well and save the keytab.

It can be as easy as:
```
addprinc -randkey HTTP/example.com
ktadd -k nifi.keytab HTTP/example.com
```

Then you need to get `nifi.keytab` add it to your volume mount and configure the `nifi.kerberos.spnego.principal` and `nifi.kerberos.spnego.keytab.location` properties and you should be ready to go.

To test it you need to use `kinit` on your local machine and configure your browser to be able to use SPNEGO. For an example see: https://bryanbende.com/development/2016/08/31/apache-nifi-1.0.0-kerberos-authentication

**Comment by SarthakSahu** [comment]
@pepov I am just trying to configure the environment for spnego testing. Once I can able to test it, will inform for review.

**Comment by SarthakSahu** [comment]
@pepov  PR is ready with spnego enhancement and updated last review comments. Please review it.

**Comment by pepov** [comment]
Sorry I don't have the time to check this right now, need to get back to this later. What did you use to setup KDC for testing?

**Comment by SarthakSahu** [comment]
I had configure my own KDCD server and followed  https://bryanbende.com/development/2016/08/31/apache-nifi-1.0.0-kerberos-authentication.
Please have a look on this.

**Comment by pepov** [comment]
sorry, I was busy lately, but will try to get to this as soon as I can

**Comment by SarthakSahu** [comment]
Can Somebody review this please.

**Comment by pepov** [comment]
looking into this now, however I beleive this branch should be rebased against master

**Comment by pepov** [comment]
I can confirm this is working for me too, please update your branch and then I'm ready to give an approve

**Comment by pepov** [comment]
I realized I already gave the approve, now even more! :) Unfortunately I can't merge, maybe @apiri or @achristianson can help us here.

File: nifi-docker/dockerhub/README.md [hunk 1]
```diff
@@ -144,6 +144,36 @@ volume to provide certificates on the host system to the container instance.
     -e LDAP_TLS_TRUSTSTORE_PASSWORD: ''
     -e LDAP_TLS_TRUSTSTORE_TYPE: ''
 
+### Standalone Instance, Kerberos
+In this configuration, the user will need to provide certificates and the associated configuration information.
+Of particular note, is the `AUTH` environment variable which is set to `kerberos`.  Additionally, the user must 
+provide a username as provided by the configured kerberos server in the `INITIAL_ADMIN_IDENTITY` environment 
+variable if wants to authenticate by kerberos. This value will be used to seed the instance with an initial user 
+with administrative privileges.  Finally, this command makes use of a volume to provide certificates on the host 
+system to the container instance.  Spnego information has to be provieded for SSO.
```
Comment (pvillard31) [inline nifi-docker/dockerhub/README.md]:
```suggestion
system to the container instance.  SPNEGO information has to be provided for SSO.
```

**Comment by achristianson** [comment]
@pepov taking a look

**Comment by achristianson** [comment]
@SarthakSahu can you squash your commits please?

**Comment by SarthakSahu** [comment]
@achristianson squash  done.

**Comment by achristianson** [comment]
Thanks @SarthakSahu. Could you take another look? It still looks like there are 15 commits on the PR. Did you force push?

**Comment by pepov** [comment]
@SarthakSahu you can check this other PR for a reference on how you can get rid of the merge commits and have everything under a single commit:
https://github.com/apache/nifi/pull/3257#issuecomment-467010401
(Please check previous and later comments too for more context)

**Comment by SarthakSahu** [comment]
I had followed the steps suggested by @pepov. @achristianson  Kindly check now and let me know if any thing wrong.

**Comment by pvillard31** [comment]
Unfortunately, even though you now have one commit, the commit is including changes not related to your PR =/

**Comment by SarthakSahu** [comment]
Is it fine now @pepov  @pvillard31  @achristianson

**Comment by MikeThomsen** [comment]
@pvillard31 if your approval still stands, I'll try to do the git work for @SarthakSahu to merge.

**Pull request closed without merging**